Sources/ServiceLifecycle/ServiceRunner.swift

```diff
@@ -0,0 +1,198 @@
+    ///   - services: The services to run.
+    ///   - configuration: The runner's configuration.
+    ///   - logger: The logger.
+    public init(
+        services: [any Service],
+        configuration: ServiceRunnerConfiguration,
+        logger: Logger
+    ) {
+        self.services = services
+        self.configuration = configuration
+        self.logger = logger
+    }
+
+    /// Runs all the services by spinning up a child task per service.
+    /// Furthermore, this method sets up the correct signal handlers
+    /// for graceful shutdown.
+    public func run(file: String = #file, line: Int = #line) async throws {
+        switch self.state {
+        case .initial:
+            self.state = .running
+            try await self._run()
+
+            switch self.state {
+            case .initial, .finished:
+                fatalError("ServiceRunner is in an invalid state \(self.state)")
+
+            case .running:
+                self.state = .finished
+            }
+
+        case .running:
+            throw ServiceRunnerError.alreadyRunning(file: file, line: line)
+
+        case .finished:
+            throw ServiceRunnerError.alreadyFinished(file: file, line: line)
+        }
+    }
+
+    private func _run() async throws {
+        self.logger.info(
+            "Starting service lifecycle",
+            metadata: [
+                self.configuration.logging.signalsKey: "\(self.configuration.gracefulShutdownSignals)",
+                self.configuration.logging.servicesKey: "\(self.services)",
+            ]
+        )
+
+        enum ChildTaskResult {
+            case serviceFinished(any Service)
+            case serviceThrew(any Service, any Error)
+            case signalCaught(UnixSignal)
+            case signalSequenceFinished
+        }
+
+        let shutdownGracefullyManager = GracefulShutdownManager()
+        try await TaskLocals.$gracefulShutdownManager.withValue(shutdownGracefullyManager) {
+            // Using a result here since we want a task group that has non-throwing child tasks
+            // but the body itself is throwing
+            let result = await withTaskGroup(of: ChildTaskResult.self, returning: Result<Void, Error>.self) { group in
+                // First we have to register our signals.
+                let unixSignals = await UnixSignalsSequence(trapping: self.configuration.gracefulShutdownSignals)
+
+                group.addTask {
+                    for await signal in unixSignals {
+                        return .signalCaught(signal)
+                    }
+
+                    return .signalSequenceFinished
+                }
+
+                for service in self.services {
+                    self.logger.debug(
+                        "Starting service",
+                        metadata: [
+                            self.configuration.logging.serviceKey: "\(service)",
+                        ]
+                    )
+
+                    group.addTask {
```

Comment:
Currently there isn't a way for a service to indicate the priority it needs to run at which seems important for certain classes of services - logging, metrics backends for example.

Reply:
Interesting point. I think priority in our ecosystem is a bit of an odd thing since a lot of services are going to be backed by NIO channels & ELs. At that point, the priority is kinda lost because these threads don't use priorities.

My current thinking is to leave priorities out for the initial release. We can add them in later through either extending the `Service` protocol or making the configuration more sophisticated. WDYT?

Reply:
We have definitely been able to improve our worst case latencies under load by considering priority. I agree with your point about priority when it gets the NIO layer but for instance our metrics backend uses an async sequence to batch and throttle the metric events before trying to send them across the network. So the original events are potentially an order of magnitude higher frequency than what is being handled by an EL.

It may come down to load testing but potentially not considering priority will block us adopting this across the board.

Reply:
I am curious to how you schedule your different services right now. Are you using a task group to run all of them?

The background for my question is that priority escalation should kick in here and actually make your lower priority service run at the higher priority again. Simple example

```swift
Task(priority: .medium) {
  await withTaskGroup(of: Void.self) { group in
    group.add(priority: .low) {
        await metricsService.run()
    }
    
    await group.next() // This will escalate the priority of the child task to `medium` 
  }
}
```

Reply:
To follow up on this. I have just checked how task priorities work on Linux and they are basically a no-op. There is no API on Linux Dispatch to implement this.

So I am surprised that you saw an improvement of latencies. Did you measure this on Linux or Darwin?